src/ser.rs

```diff
@@ -0,0 +1,1130 @@
+                                -> Result<Self::SerializeStructVariant, Self::Error> {
+        Err(Error::DateInvalid)
+    }
+}
+
+#[doc(hidden)]
+pub struct Impossible<T> {
+    _inner: T,
+    nope: Nope,
+}
+
+enum Nope {}
+
+impl<T> ser::SerializeSeq for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_element<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeMap for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_key<U: ?Sized>(&mut self, _key: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn serialize_value<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeStruct for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _key: &'static str, _value: &U)
+                                  -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeTuple for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_element<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeTupleStruct for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeTupleVariant for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeStructVariant for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _key: &'static str, _value: &U)
+                                  -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+struct StringExtractor<'a> {
+    slot: &'a mut String,
+}
+
+impl<'a> ser::Serializer for StringExtractor<'a> {
+    type Ok = ();
```

Comment:
I think this would be clearer if StringExtractor is a unit struct and StringExtractor::Ok = String.